[SPARK-8581][SPARK-8584] Simplify checkpointing code + better error message

andrewor14 opened the pull request:
This patch rewrites the old checkpointing code in a way that is easier to understand. It also adds a guard against an invalid specification of checkpoint directory to provide a clearer error message. Most of the changes here are relatively minor.

Commit:
Simplify checkpoint code / code style / comments

Commit:
Guard against bad settings of checkpoint directory

andrewor14 commented:
@tdas PTAL

andrewor14 commented on core/src/main/scala/org/apache/spark/rdd/CheckpointRDD.scala:
```diff
@@ -59,9 +61,6 @@ class CheckpointRDD[T: ClassTag](sc: SparkContext, val checkpointPath: String)
   }
 
-  checkpointData = Some(new RDDCheckpointData[T](this))
-  checkpointData.get.cpFile = Some(checkpointPath)
```
I removed this because the checkpoint data here isn't really used; it's only there for `getCheckpointFile` to return the right path, which we can just provide by overriding it ourselves.

andrewor14 commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -1897,6 +1897,13 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
+    if (!isLocal && Utils.nonLocalPaths(directory).isEmpty) {
+      throw new IllegalArgumentException(
+        "Checkpoint directory must be non-local if Spark is running on a cluster: " + directory)
+    }
```
This should be the only real change in functionality.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35621 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35621/consoleFull) for   PR 6968 at commit [`7fb4af5`](https://github.com/apache/spark/commit/7fb4af546c7d6d7d78238ba633ae1c744e805061).

SparkQA commented:
[Test build #35621 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35621/console) for   PR 6968 at commit [`7fb4af5`](https://github.com/apache/spark/commit/7fb4af546c7d6d7d78238ba633ae1c744e805061).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

andrewor14 commented:
retest this please??

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35623 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35623/consoleFull) for   PR 6968 at commit [`7fb4af5`](https://github.com/apache/spark/commit/7fb4af546c7d6d7d78238ba633ae1c744e805061).

SparkQA commented:
[Test build #35623 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35623/console) for   PR 6968 at commit [`7fb4af5`](https://github.com/apache/spark/commit/7fb4af546c7d6d7d78238ba633ae1c744e805061).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

JoshRosen commented:
What if I have an NFS filesystem that's mounted in a way that it appears as a local disk?  I wonder if we should have a flag to bypass the non-local check for those rare edge-cases.

andrewor14 commented:
retest this please

andrewor14 commented:
Hm that seems hard to detect. Maybe we should just do away with the check.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35704 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35704/consoleFull) for   PR 6968 at commit [`7fb4af5`](https://github.com/apache/spark/commit/7fb4af546c7d6d7d78238ba633ae1c744e805061).

Commit:
Merge branch 'master' of github.com:apache/spark into checkpoint-cleanup

Commit:
Warn instead of throw

... in case the user provides a NFS mounted directory.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35713 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35713/consoleFull) for   PR 6968 at commit [`b1437ad`](https://github.com/apache/spark/commit/b1437ada6cb251e1ccc8f2aa896df0f53f8b51f7).

SparkQA commented:
[Test build #35704 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35704/console) for   PR 6968 at commit [`7fb4af5`](https://github.com/apache/spark/commit/7fb4af546c7d6d7d78238ba633ae1c744e805061).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

SparkQA commented:
[Test build #35713 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35713/console) for   PR 6968 at commit [`b1437ad`](https://github.com/apache/spark/commit/b1437ada6cb251e1ccc8f2aa896df0f53f8b51f7).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

andrewor14 commented:
retest this please

andrewor14 commented:
@tdas can you review?

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36006 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36006/consoleFull) for   PR 6968 at commit [`b1437ad`](https://github.com/apache/spark/commit/b1437ada6cb251e1ccc8f2aa896df0f53f8b51f7).

SparkQA commented:
[Test build #36006 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36006/console) for   PR 6968 at commit [`b1437ad`](https://github.com/apache/spark/commit/b1437ada6cb251e1ccc8f2aa896df0f53f8b51f7).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

tdas commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -1897,6 +1897,16 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
+    // its own local file system, which is incorrect because the checkpoint files
+    // are actually on the executor machines.
+    if (!isLocal && Utils.nonLocalPaths(directory).isEmpty) {
+      logWarning("Checkpoint directory must be non-local " +
```
awesome!

tdas commented on core/src/main/scala/org/apache/spark/rdd/RDD.scala:
```diff
@@ -194,7 +194,7 @@ abstract class RDD[T: ClassTag](
 
   /** An Option holding our checkpoint RDD, if we are checkpointed */
-  private def checkpointRDD: Option[RDD[T]] = checkpointData.flatMap(_.checkpointRDD)
+  private def checkpointRDD: Option[CheckpointRDD[T]] = checkpointData.flatMap(_.checkpointRDD)
```
Either is okay, but hardcoding the type here seems unnecessary. Same reason we use Seq[T] in a lot of places rather than Array[T].

andrewor14 replied:
how is this hardcoding? Semantically we want the type of this to always be `CheckpointRDD`. This is just enforcing it in code.

tdas replied:
As i said either is okay by me. Just a thought that if we change the type name of CheckpointRDD, or have different kinds of CheckpointRDDs depending on the RDD, then this code will not require any change. RDD[T] is necessary and sufficient. 

Again, minor point.

andrewor14 replied:
I disagree. It's bad practice to have super general types everywhere. If we want different kinds of `CheckpointRDD`s we should have a common `CheckpointRDD` abstract class and plug in arbitrary implementations.

squito replied:
fwiw, I agree with Andrew

tdas replied:
I agree that the ideal choice out here is defining a trait that clearly narrows down what can be assigned to it.

tdas commented on core/src/main/scala/org/apache/spark/rdd/RDD.scala:
```diff
@@ -194,7 +194,7 @@ abstract class RDD[T: ClassTag](
-  private def checkpointRDD: Option[RDD[T]] = checkpointData.flatMap(_.checkpointRDD)
+  private def checkpointRDD: Option[CheckpointRDD[T]] = checkpointData.flatMap(_.checkpointRDD)
 
   /**
```
Since you are cleaning up the code, may i suggest another change. It is probably a good idea to mark RDD.checkpointData as volatile. Its a var, and assigning to it in `def checkpoint()` is not synchronized. Can cause race conditions.

andrewor14 replied:
ok, I did this.

tdas commented on core/src/main/scala/org/apache/spark/rdd/RDDCheckpointData.scala:
```diff
@@ -46,37 +45,35 @@ private[spark] class RDDCheckpointData[T: ClassTag](@transient rdd: RDD[T])
-  def isCheckpointed: Boolean = {
-    RDDCheckpointData.synchronized { cpState == Checkpointed }
+  def isCheckpointed: Boolean = synchronized {
+    cpState == Checkpointed
```
Offline conversation: Global lock necessary to ensure correctness.

tdas replied:
Could you add the reasoning in the documentation.

andrewor14 replied:
offline conversation: no one knows why it has to be a global lock, so no documentation needed!

Commit:
Merge branch 'master' of github.com:apache/spark into checkpoint-cleanup

Commit:
Use global synchronized instead

Per @tdas in offline discussions. Let's just try to keep this
patch as simple as possible.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36318 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36318/consoleFull) for   PR 6968 at commit [`4ef8263`](https://github.com/apache/spark/commit/4ef826368e527fb90739aade0902fd0aecf6dcbd).

tdas commented:
LGTM. :)

SparkQA commented:
[Test build #36318 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36318/console) for   PR 6968 at commit [`4ef8263`](https://github.com/apache/spark/commit/4ef826368e527fb90739aade0902fd0aecf6dcbd).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

andrewor14 commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36348 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36348/consoleFull) for   PR 6968 at commit [`4ef8263`](https://github.com/apache/spark/commit/4ef826368e527fb90739aade0902fd0aecf6dcbd).

SparkQA commented:
[Test build #36348 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36348/console) for   PR 6968 at commit [`4ef8263`](https://github.com/apache/spark/commit/4ef826368e527fb90739aade0902fd0aecf6dcbd).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

andrewor14 commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36405 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36405/consoleFull) for   PR 6968 at commit [`4ef8263`](https://github.com/apache/spark/commit/4ef826368e527fb90739aade0902fd0aecf6dcbd).

SparkQA commented:
[Test build #36405 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36405/console) for   PR 6968 at commit [`4ef8263`](https://github.com/apache/spark/commit/4ef826368e527fb90739aade0902fd0aecf6dcbd).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

andrewor14 commented:
Merging into master.
